packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts

```diff
@@ -0,0 +1,235 @@
+import { LaunchRoleProps, ProvisioningRulesProps, EventNotificationsProps, TagUpdatesProps, StackSetConstraintProps, ConstraintProps } from '../constraints';
+import { IPortfolio, Portfolio } from '../portfolio';
+import { IProduct } from '../product';
+import { CfnPortfolioProductAssociation, CfnLaunchRoleConstraint, CfnLaunchTemplateConstraint, CfnLaunchNotificationConstraint, CfnResourceUpdateConstraint, CfnStackSetConstraint, CfnTagOption, CfnTagOptionAssociation } from '../servicecatalog.generated';
+import { getIdentifier } from './util';
+import { InputValidator } from './validation';
+
+export class AssociationManager {
+
+  public static associateProductWithPortfolio(scope: Resource, portfolio: IPortfolio, product: IProduct) {
+    const associationKey = getIdentifier(portfolio.node.addr, product.node.addr, scope.stack.node.addr);
+    if (!this.associationMap.has(associationKey)) {
+      const association = new CfnPortfolioProductAssociation(scope, `PortfolioProductAssociation${associationKey}`, {
+        portfolioId: portfolio.portfolioId,
+        productId: product.productId,
+      });
+      this.associationMap.set(associationKey, new Map());
+      this.associationMap.get(associationKey).set(Constraints.ASSOCIATION, association);
+    }
+  }
+
+  public static addLaunchRoleConstraint(scope: Resource, props: LaunchRoleProps) {
+    const [pair, associationKey] = this.associationPrecheck(scope, props);
+    InputValidator.validateLength(this.generateAssocationString(scope, pair), 'description', 0, 2000, props.description);
+    if (!(this.associationMap.get(associationKey).get(Constraints.LAUNCH_ROLE)) &&
+     !(this.associationMap.get(associationKey).get(Constraints.STACKSET))) {
+      const constraint = new CfnLaunchRoleConstraint(scope, `LaunchRoleConstraint${associationKey}`, {
+        acceptLanguage: props.acceptLanguage ?? AcceptLanguage.EN,
+        description: props.description ?? '',
+        portfolioId: pair.portfolio.portfolioId,
+        productId: pair.product.productId,
+        roleArn: props.role.roleArn,
+      });
+
+      constraint.addDependsOn(this.associationMap.get(associationKey).get(Constraints.ASSOCIATION));
+      this.associationMap.get(associationKey).set(Constraints.LAUNCH_ROLE, constraint);
+    } else {
+      throw new Error(`Cannot have multiple launch or stackset constraints on association ${this.generateAssocationString(scope, pair)}`);
+    }
+  }
+
+  public static addTemplateConstraint(scope: Resource, props: ProvisioningRulesProps) {
+    const [pair, associationKey] = this.associationPrecheck(scope, props);
+    InputValidator.validateLength(this.generateAssocationString(scope, pair), 'description', 0, 2000, props.description);
+    const rules = JSON.stringify(
+      {
+        Rules: props.rules,
+      },
+    );
+    if (rules == '{}') {
+      throw new Error(`No rules provided for provisioning for association ${this.generateAssocationString(scope, pair)}`);
+    }
```

Comment:
Uhm. Basing this check on a string serialization doesn't seem great (what if it returns `{}\n`?).

How about checking whether `props.rules` is undefined and/or empty instead?